### ISSUE TITLE
More fine-grained control over user folders/albums

### ISSUE THREAD

**Issue by ScuttleSE** [photoview/photoview#416]
As far as I see it now, the user-management and separation of content between users is 100% dependent on the actual folder structure of your photo library.

There is no way to select certain folders/albums to give a user access to. If a user gets access to /photos/Boats, it will get access to all Albums that exists in /photos/Boats and its subfolders.

A more granular way to give users access to albums would be appreciated. Something that is not dependent on the physical structure of the filesystem.

**Comment by viktorstrate** [comment]
Some have suggested a "virtual album" feature similar to collections in Lightroom, that could potentially be part of a solution. Alternativeley if users not only had the option to share albums with a public link but also with other users directly then that could be a solution as well.

**Comment by Redsandro** [comment]
I would love to easily give different users access to different albums, as well as choosing which users are allowed to create shared albums (#435) and share the "people" labeled faces within groups of users so they can collaborate (#436).

**Comment by Sieboldianus** [comment]
I agree, having a way to permanently share specific folders, collections etc. with specific users would be great. For example, I want to give automatic access to familiy member to all subfolders created in `/photos/me/nextcloud/InstantPhotoUpload/` - currently, there is not a neat way to do this. Users must memorize the shared link. In other Apps, usually this is implemented by having a subfolder "Shared with me" (e.g. see GDrive), where those Shares are listed.

**Comment by viktorstrate** [comment]
@Sieboldianus thank you for your donation ❤️

@Sieboldianus + @Redsandro:
You can add the same path to multiple users, and everyone will have access to them and the server will not generate duplicate thumbnails etc. Also faces in images shared by multiple users only exist as one, this means that if one user merges two faces both users see the change. I think this means that there is no reason to have groups right?

I still think it would be nice if users could share albums with each other, on a non-admin level from the sidebar, and add a "Shared with me" like menu item.

**Comment by Redsandro** [comment]
@viktorstrate commented:  

> You can add the same path to multiple users, and (...) faces in images shared by multiple users only exist as one, this means that if one user merges two faces both users see the change.  
  
I double-checked and in the end it is like that indeed. But it takes a couple of hours. :thinking: Is everything scanned again before it shows up? Can we skip that, since it's the same path and the metadata already exists? I want to be able to add you (for example) to my instance of photoview and you'll have access to the faces and albums immediately. Not after a couple of hours of scanning.

> I think this means that there is no reason to have groups right?

No I still think groups is missing. Imagine a group with 20 members, and you want to add another path. Now you have to make the same change 20 times. And as mentioned above, if someone messages you and asked "hey can I have access too" you want all the data that has already been crawled to be available to them immediately.

**Comment by Sieboldianus** [comment]
> I still think it would be nice if users could share albums with each other, on a non-admin level from the sidebar, and add a  "Shared with me" like menu item.

Just to add to the discussion, I think how Flickr solved this is nice: Each picture can have 3 types of permissions assigned: public, private, and "family" - the latter those connected users checked as a family member to view an intermediate set of semi-private pictures. This is very simple but powerful, since everybode can decide what "family" means. It also works on a per-file basis, making it quite flexible. In photoview, I could imagine that you could browse another users photo the same way as browsing ones one collection (Photos/Albums/Maps/Faces), just that they are filtered to the subset of photos that are either public, directly shared with you, or visible because you are a "family member".

I know - this is smehow overstretching the discussion here.

**Comment by try-and-succeed** [comment]
I'd like to work on this and put a design up for discussion before writing code. It came out of a larger branch (#1473, since split up and closed) that tried per-album grants and kept running into consistency bugs.

### What went wrong with stored grants

The branch stored who can see which album as rows: one per user and album, copied down to every descendant. Each operation then had to keep those rows right: moving a folder, a scan finding new sub-folders, revoking a share, removing a root path. Ten review rounds kept finding cases where the stored rows and the real tree disagreed, and where they did, someone saw too much or too little.

### Proposal: compute visibility, don't store it

- **Two concepts, kept apart:**
  - **Ownership** is what exists today. An admin gives a user a root path, and the user has full rights below it. Unchanged.
  - **A share** gives another user a specific album. **Shares are read-only**: the recipient can view and download, but not upload, delete, move or re-share.
- **Only the share itself is stored**: `(album, user)`. Whether a user can see an album is computed: *they own a root above it, or some album above it (or the album itself) is shared with them.*
- Album paths are already filesystem paths, and an album's path starts with its parent's. So "is some ancestor shared with me" is a **path-prefix test** against the handful of albums shared with that user. No recursive query and no copied rows. A folder moved into or out of a shared folder is visible or not *by construction*, with no bookkeeping step that could be forgotten.

That covers the requests in this thread. A family member can get `/photos/me/nextcloud/InstantPhotoUpload/` shared once and sees every folder created there later, with no per-folder action and no share link to remember.

### Things to get right

- The prefix test needs a trailing separator, so `/photos/anna` does not match `/photos/anna2`.
- Albums reached through a symlink store the path the scan walked, which can differ from the resolved one. That needs a decision on which path counts.
- Existing installs: today's rows are ownership only. A migration would have nothing to convert, but it needs checking against the real schema.
- UI: a "Shared with" section in the album sidebar (owner only), and "shared with me" albums shown separately from the user's own.

### Questions for the maintainers

1. Is read-only sharing enough for a first version, or do you see a need for "can upload" shares? It could come later as a flag on the share.
2. Should shared albums appear in the recipient's timeline and search, or only under "Shared with me" at first?
3. Would you rather have this as one PR (model + UI), or the model and its GraphQL API first, the UI second?

If the direction sounds right, I'll start with the model and its tests.

**Comment by kkovaletp** [comment]
@try-and-succeed, thanks for sharing your vision and being willing to contribute a solution!

In general, I'd prefer having virtual collections feature implemented 1st and this feature being added on top of collections: this approach would give much more flexibility to users, as we're not limited with FS structure, sub-albums, etc. Users would be able to create collections with any structure inside and add the same media to as many collections as they need. Then, sharing such collections with other users would be as easy as adding other users to those collections, with an optional permission for resharing (off by default).

Implementing this feature for the existing albums structure would add more mess to an already complex enough multi-user FS structure, and I'd expect more corner cases to be uncovered...

Answers to your questions:
1. Photoview is read-only. What do you mean by "can upload" shares? The only question here if we want users to control the re-sharing capability for shared albums/collections. For the 1st version, we can go with non-re-sharable albums/collections. But it makes sense to design the architecture in a way allowing to add the "can re-share" capability later.
2. I'd say they should appear everywhere in the app for consistency (timeline, search, map, people), but it should be easily distinguished from the user's owned media (with a "shared" overlay/icon/tag, or something similar)
3. The best would be to split this feature into an ordered list of smaller, fully functional chunks that can be delivered 1-by-1 and add the corresponding functionality in a usable way without breaking the existing functionality in `master`

**Comment by try-and-succeed** [comment]
@kkovaletp, thanks. Two things.

**Sharing.** My need is small and concrete: a family member gives another user read-only access to one folder. Virtual collections are a much bigger feature, and it may take a long time until someone builds them. I'd rather have a small solution now than live without one for a long time. To be clear, I don't mean the big feature should be built on top of my code — it can be implemented however it makes sense. I only mean that whatever collections bring, sharing a folder read-only will be part of what they can do, so the small step doesn't lead anywhere that has to be abandoned.

**Uploading.** You say Photoview is read-only, and I know #129 was closed on that principle back in 2020. That principle is the one thing that stops Photoview from solving my actual problem: my daughter's phone runs out of space, and she should be able to move a folder into her own library area from the phone and free that space. The answer given back then — use FTP or a sync app and let Photoview pick the files up — works, but it means another service, another set of accounts and passwords, and explaining all of it to the family. With an upload in Photoview itself, everything comes from one application with the account they already have, which is what would make it genuinely comfortable to use.

Would an optional upload be acceptable in principle — off by default, enabled per user by an admin, writing only into that user's own folder? If yes, I'd propose it as its own issue later. If not, that is a clear answer and it stays in my fork.

Either way I won't start anything before my open PRs are merged or closed. Your review time is the limited resource here, so I'd rather not have too much running in parallel.

**Comment by kkovaletp** [comment]
@try-and-succeed, here are my answers to your comments:

> **Sharing.** My need is small and concrete: a family member gives another user read-only access to one folder.

While your use case is focused and limited in scope, every new feature has to be consistent with the current architecture and functionality, as well as address standard user needs (so others benefit from it). Adding the sharing functionality on top of Albums in a consistent way, addressing standard user workflows, would quickly make the PR huge because of many corner cases and other things. Also, there will be a lot of conceptual questions about how to handle different cases (like symlinks, sub-albums, recognized people,...) The reason is that the current architecture doesn't expect this functionality and is not ready. That is why I think that creating a new media structure "Collections" in the app and designing its architecture and functionality from scratch would be a smoother flow.

> **Uploading.** You say Photoview is read-only, and I know [#129](https://github.com/photoview/photoview/issues/129) was closed on that principle back in 2020.

Still the same answer. Photoview is designed to be a read-only media gallery, and it will stay read-only to the source media FS structure. We have a list of feature requests in the backlog that require some data modification (like collections, which would allow adding some media to them from inside Photoview), but all these modifications are limited to Photoview DB and metadata of the source media files - nothing is going to be modified in the source media location.
There are many reasons for that: the product architecture, huge security risks of letting the web service modify users' media library, and more...
So, it will stay read-only to the source media location(s).

**Comment by try-and-succeed** [comment]
Thanks for the clear answers — that settles both questions for me.

Uploading stays in my fork then. I understand the reasoning.

On sharing: understood, I won't propose it on top of albums. I'll keep my folder-level version in my fork and leave this issue to the collections design.

That leaves my eight open PRs, none of which touch any of this. I'll keep to those.
